File: Python/Spiral Matrix.py
"""
Given a matrix of m x n elements (m rows, n columns), return all elements of the matrix in spiral order.

For example,
Given the following matrix:

[
 [ 1, 2, 3 ],
 [ 4, 5, 6 ],
 [ 7, 8, 9 ]
]
You should return [1,2,3,6,9,8,7,4,5].
"""

class Solution:
    # @param matrix, a list of lists of integers
    # @return a list of integers
    def spiralOrder(self, matrix):
        result = []
        if matrix == [] or matrix[0] == []:
            return result
        left, right, top, bottom = 0, len(matrix[0]) - 1, 0, len(matrix) - 1
        while left <= right and top <= bottom:
            for i in range(left, right + 1):
                result.append(matrix[top][i])
            for i in range(top + 1, bottom):
                result.append(matrix[i][right])
            for i in reversed(range(left, right + 1)):
                if top < bottom:
                    result.append(matrix[bottom][i])
            for i in reversed(range(top + 1, bottom)):
                if left < right:
                    result.append(matrix[i][left])
            left, right, top, bottom = left + 1, right - 1, top + 1, bottom - 1
        return result

File: Java/Spiral Matrix.java
/*
Given a matrix of m x n elements (m rows, n columns), return all elements of the matrix in spiral order.

For example,
Given the following matrix:

[
 [ 1, 2, 3 ],
 [ 4, 5, 6 ],
 [ 7, 8, 9 ]
]
You should return [1,2,3,6,9,8,7,4,5].
*/

public class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        ArrayList<Integer>result = new ArrayList<>();
        if(matrix.length == 0 || matrix[0].length == 0) return result;
        int left = 0,right = matrix[0].length-1,top = 0,bottom = matrix.length-1;
        while(left <= right && top <= bottom){
            for(int i=left;i<right+1;i++)
                result.add(matrix[top][i]);
            for(int i=top+1;i<bottom;i++)
                result.add(matrix[i][right]);
            for(int i=right;i>=left;i--){
                if(top<bottom)
                    result.add(matrix[bottom][i]);
            }
            for(int i=bottom-1;i>=top+1;i--){
                if(left<right)
                    result.add(matrix[i][left]);
            }
            left++;right--;top++;bottom--;
        }
        return result;
    }
}

public class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> elements = new ArrayList<>();
        if (matrix.length == 0) return elements;
        int m = matrix.length, n = matrix[0].length;
        int row = 0, col = -1;
        while (true) {
            for (int i = 0; i < n; i++) {
                elements.add(matrix[row][++col]);
            }
            if (--m == 0) break;
            for (int i = 0; i < m; i++) {
                elements.add(matrix[++row][col]);
            }
            if (--n == 0) break;
            for (int i = 0; i < n; i++) {
                elements.add(matrix[row][--col]);
            }
            if (--m == 0) break;
            for (int i = 0; i < m; i++) {
                elements.add(matrix[--row][col]);
            }
            if (--n == 0) break;
        }
        return elements;
    }
}